core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -242,6 +244,14 @@ class LogSegment(val log: FileRecords,
           lastIndexEntry = validBytes
         }
         validBytes += batch.sizeInBytes()
+
+        if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
+          leaderEpochCache match {
+            case Some(epochCache) =>
+              epochCache.assign(batch.partitionLeaderEpoch, batch.baseOffset())
```

Comment:
It think we need the following check to avoid the unnecessary warn logging in maybeWarn during assign().

if (batch.partitionLeaderEpoch > epochCache.latestEpoch())